### DISCUSSION TITLE
Should a `joinedload()` option remain attached to an instance beyond the next `commit()`?

### DISCUSSION THREAD

**Discussion by rsyring** [sqlalchemy/sqlalchemy#12030] [Usage Questions] [answered]
Big picture, we have a mixin class that provides a `fake()` method on our Entities that creates instances with faked data.  That mixin creates the instance, adds it to the session, and then commits.  It's often the case that our tests will do multiple fakes in succession and then make queries, potentially commit, and then the entities will be tested to see if they are updated correctly.

This obviously depends on SQLAlchemy's refresh-after-commit functionality to work correctly.

I've also been trying to move to an explicit load requirement in our code base to force me/us to think about this aspect of SA for every query we write since it can easily be the source of performance problems.  Therefore, I've started using `lazy=raise` on relationships and then providing a base query method on the entity that is more explicit & intentional about how relationships are joined and loaded.

However, I want my faked instances to be more flexible and so I'm looking for a way to configure the faked instances so that any relationship works and I don't care so much about what loading strategy is used.  Although, the method below made sense to me (i.e. just eager load everything for a faked instance()).

Some pseudo code which is a minimized  overview of what I'm trying to do.  Note this code is not intended to run as-is since I'm piecing together a few things that are currently buried in libraries:

```python
import sqlalchemy as sa
from sqlalchemy import orm
from flask_sqlalchemy import SQLAlchemy

db = SQLAlchemy()

class Project(db.Model): ...


class Task(db.Model):
    __tablename__ = 'task'

    project_id = sa.Column(sa.String, sa.ForeignKey(Project.id, ondelete='cascade'), nullable=False)
    project = orm.relationship(Project, lazy='raise', load_on_pending=True)

    @classmethod
    def fake(cls):
        task = cls(**fake_task())
        db.session.add(task)
        db.session.commit()
        return cls.q_join().where(cls.id == task.id).one()

    @classmethod
    def q_join(cls):
        query = cls.query.join(cls.project).options(
            orm.joinedload(cls.project),
        )

        return query
    
task = Task.fake()

# Without this second fake, there is no error.
Task.fake()

# Raises: 'Task.project' is not available due to lazy='raise'
assert task.project.name == '...'
```

And the relevant part of the exception:

```pytb
  File "/home/rsyring/projects/kilo-src/apps/doist-pkg/doist/tests/test_entities.py", line 426, in test_q_join
    assert task.project.name
  File "/home/rsyring/.local/pipx/venvs/doist/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py", line 566, in __get__
    return self.impl.get(state, dict_)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rsyring/.local/pipx/venvs/doist/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py", line 1086, in get
    value = self._fire_loader_callables(state, key, passive)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rsyring/.local/pipx/venvs/doist/lib/python3.11/site-packages/sqlalchemy/orm/attributes.py", line 1121, in _fire_loader_callables
    return self.callable_(state, passive)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rsyring/.local/pipx/venvs/doist/lib/python3.11/site-packages/sqlalchemy/orm/strategies.py", line 915, in _load_for_state
    self._invoke_raise_load(state, passive, "raise")
  File "/home/rsyring/.local/pipx/venvs/doist/lib/python3.11/site-packages/sqlalchemy/orm/strategies.py", line 867, in _invoke_raise_load
    raise sa_exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: 'Task.project' is not available due to lazy='raise'
```

The second `Task.fake()`, due to the `commit()` in `fake()`, results in an expiration and reloading of `task` the next time it's used.  I'm guessing this is the source of the problem.  But that task was originally created from query with a `joinedload()` option set and I was under the impression that option was supposed to remain "sticky" on the instance for when it was reloaded/refreshed in the future?

I can troubleshoot further if needed but wanted to make sure I didn't have a mistake in my mental model of how this is supposed to work.

I'd also be open to other suggestions on how to use lazy='raise' on the relationship while also creating some instances that function as if `lazy=joined` was the relationship default for as long as they exist in the session, regardless of expiration / refresh.

**Comment by zzzeek** [comment, marked as answer]
joinedload() is sticky if it takes effect for that entity.  in your example, you have several things going on preventing it from doing what you want it to do.

first is that your query inside of fake() isn't actually loading the object.   Because you have called commit(), all objects are expired.  then you access `task.id`.  that **refreshes the object fully at that moment**.  Then you query for that same object, a SELECT is emitted, but the state of the object  is not touched.   You would need to either a. dont refresh the object via `.id` after the commit or b. use populate_existing (b is more reliable):

```
        return (
            cls.q_join()
            .where(cls.id == task.id)
            .execution_options(populate_existing=True)
            .one()
        )

```

with that option in place, we can see the joinedload() option is in fact sticky:

```
task = Task.fake()

breakpoint()

(Pdb) from sqlalchemy import inspect
(Pdb) inspect(task).load_options
(<sqlalchemy.orm.strategy_options.Load object at 0x7fdf1f4ae700>,)
```


but after that we get into an unfortunate chicken/egg scenario for which I dont have a clean solution right now.   it's only in recent SQLAlchemy versions that an expired object will fire off eager loaders when it unexpires, so this is a new functionality.  however it does not work in this case because you are accessing .project directly without touching any other attributes.   this fires off the "raiseload" loader you have installed before the lazy loader attempts to look at the foreign key attribute on the object; an attempt which would unexpire the object and end up running the joinedload.    it would be a very big change right now to try to change this order of operations and then to enhance the lazyloader to realize that the joinedload() already took place, I'd label this as a quagmire issue.  it is definitely not 2.0 and im not sure how feasible it is in general, it would be a long term change for future SQLAlchemy versions but it's not clear if there are reasons it can't actually work.

if you access any columns on that task, or run session.refresh(), then the `.project` attribute is there.

The larger pattern causing this to fail is that you are calling .commit(), with expire_on_commit, in the middle of a series of object operations where you are hoping to have consistency between them, if this is a test suite there's likely no chance of out-of-band data from other transactions coming in so I'd use flush() here or turn off expire_on_commit.

**Reply by rsyring** [reply]
Really, really appreciate the thorough explanation!  Brings a lot of clarity and plenty to work out some kind of solution.  Thank you.  :)
